implement array operations methods and simple login system

gregory-butler opened the pull request:
I implemented methods for each array operation and stored them in their own static class named 'Arrays'. For each of the method in the 'Arrays' class I created an correspondent in the 'Demo' class to demonstrate and test that the methods work correctly.
While I was working on this I discovered how to create XML comments and wrote them for the 'Arrays' class.
For the last part of the homework I created another project called 'LoginSimulator' in which I implemented two methods: one to prompt user for the name and password without showing pressed keys and the other to test credential against stored one.
LE:
I updated the logic for the 'Sort' method to use 'Merge Sort' algorithm, before I used 'Bubble Sort'. Like with the other methods I found the need to test it with different data and in different scenarios so I decided to delve a bit in to unit testing. 
The framework I chose is xUnit. 
LELE: found out that triple A is not just for battery's 😄

gregory-butler committed:
update to include config dirs for rider and vscode.

gregory-butler committed:
add initial project.

gregory-butler committed:
create a static class for the array operations part of the homework;
implement a method for each of the array operations;
implement a helper method to print the contents of an array.

gregory-butler committed:
implement methods to test each array operations functionality.

gregory-butler committed:
move test methods to their own class.

gregory-butler committed:
implement 'Pause' method and call it after array operations test methods.

gregory-butler committed:
simplify method name;
add exit message to the last call of 'Pause' method.

gregory-butler committed:
add XML documentations comments.

gregory-butler committed:
rename 'Tests.cs' back to 'Demo.cs'.

gregory-butler committed:
simplify array to string conversion.

gregory-butler committed:
add a new project for the login part.

gregory-butler committed:
implement 'PromtString' method that reads user input without key echo;
implement 'TestCredentials' method test user input against stored credential.

gregory-butler committed:
add empty line at the end of the file.

gregory-butler committed:
add exceptions to 'RemoveLast' method;
add more details the comments.

gregory-butler committed:
rename methods to better describe what they do.

gregory-butler committed:
rename 'Sort' method to be in line with the rest of the method names;
add optional argument to the 'Print' method.

gregory-butler committed:
downgrade target framework to netcore3.1;

gregory-butler committed:
update exception filther to support framework downgrade.

gregory-butler committed:
dissable nullable reference type.

gregory-butler committed:
implement 'Merge Sort' algorithm.

gregory-butler committed:
remove unused code;
'SwapElements' method was no longer used after removing 'Bubble Sort' method.

gregory-butler committed:
add comments to the 'Sort' method.

gregory-butler committed:
add xUnit project.

gregory-butler committed:
implement 'Sort' method tests.

gregory-butler committed:
add 'InsertAtTests' class.

gregory-butler committed:
beautify 'Print' methoud output.

gregory-butler committed:
refactor code to respect AAA.

gregory-butler committed:
implement tests for 'InsetAt' method.

gregory-butler committed:
update variable names;
update inline data.

gregory-butler committed:
rename 'ACT' variable.

gregory-butler committed:
remove unused argument.

gregory-butler committed:
at test to cover 'ArgumentOutOfRangeException'.

gregory-butler committed:
implement 'RemoveAt' tests.

westonpace commented on ArrayOperationTests/InsertAtTests.cs:
```diff
@@ -0,0 +1,49 @@
+  public class InsertAtTests
+  {
+    [Fact]
+    public void InsertAt_Should_Throw_ArgumentException_When_Input_Array_Is_Empty()
```
Why?  Why can't I insert into an empty array at index 0?

gregory-butler replied:
Makes a lot of sense when you are asking, the user should be able to insert an item in an empty array.

westonpace commented on ArrayOperationTests/InsertAtTests.cs:
```diff
@@ -0,0 +1,49 @@
+    { 
+      int[] actual; 
+       
+      actual = Arrays.InsertAt(input, number, index);
```
You could combine the declaration and the call here...
`int[] actual = Arrays.InsertAt(input, number, index);`

gregory-butler replied:
I only split the initialization from declaration after I read about `Arrange, Act Assert`. I can see now the error in my logic as the 'Arrange' part is done using `InlineData`.

westonpace commented on ArrayOperationTests/SortTests.cs:
```diff
@@ -0,0 +1,57 @@
+      
+      Action actual = () => Arrays.Sort(input);
+      
+      Assert.Throws<ArgumentException>(actual);
```
Again, why?  Isn't an empty array sorted?

westonpace commented on ArrayOperationTests/SortTests.cs:
```diff
@@ -0,0 +1,57 @@
+    [InlineData(new[] {1, 0, 2, -1, 3}, new[] {-1, 0, 1, 2, 3})]
+    [InlineData(new[] {1, -1, 1}, new[] {-1, 1, 1})]
+    [InlineData(new[] {1, 1, 1}, new[] {1, 1, 1})]
+    public void Sort_Should_Return_Expected_Array(int[] input, int[] expected)
```
Not a review comment so much but something to make you think.

What you're doing here is very common.  You supply the input and the expected output.

Another way to do this kind of testing is to supply an input and a verification algorithm.  For example, for a sorting algorithm you can verify by taking the result and making sure each item is >= the previous.  Then you don't have to supply the expected output.

Anyways, what you have is probably best.  I'm just rambling.

gregory-butler replied:
It's definitely not easy to read. I don't like how the inline data looks, even more so for the other tests that have more arguments. I saw there other ways to do it but it requires more OOP knowledge than I currently poses. One thing I've learned is that it's not a good idea to introduce logic in the actual tests.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+    /// <summary>
+    /// Sort an array using 'Bubble Sort' algorithm.
+    /// </summary>
+    /// <param name="input">Source array.</param>
```
If you're going to do this commenting (you don't need to) then you should also include `<returns>`

gregory-butler replied:
I just forgot to add it. I initially implemented bubble sort to do the sorting in place without returning another array.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+  public static class Arrays
+  {
+    /// <summary>
+    /// Sort an array using 'Bubble Sort' algorithm.
```
I'm pretty sure you're not using bubble sort anymore.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+      var midPoint = input.Length / 2;  
+      
+      var left = new int[midPoint];
+      var right = input.Length % 2 == 0 ? new int[midPoint] : new int[midPoint + 1];
```
This is great and it utilizes what was taught in class more effectively.  I'm only bringing this up since you are clearly putting in a lot of effort here.  For a merge sort on very large arrays, it's ideal not to have to do all these new arrays and copies.  Making changes in-place will have better performance.  If you want to make one copy at the top-level you can do that too and then you preserve the input array which is often good.

gregory-butler replied:
I understand, definitely something to study in the following days.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+          indexLeft++;
+          indexOutput++;
+        }
+        else if (indexRight < right.Length)
```
Technically this could just be `else` I think.

gregory-butler replied:
I definitely want to become as experienced as you are to notice these small things. Thank you 🤗

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+    /// <summary>
+    /// Inserts a new element in an array at the specified position.
+    /// </summary>
+    /// <param name="input">Destination array.</param>
```
I wouldn't use the word `Destination`.  The input array is going to be unmodified.  Instead you are returning a new array.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+    /// <exception cref="ArgumentException">The destination array has no elements.</exception>
+    public static int[] InsertFirst(int[] input, int number)
+    {
+      if (input == null)
```
These checks are redundant.  You will catch this in `InsertAt`.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+      }
+      
+      const int index = 0;
+      var output = InsertAt(input, number, index);
```
Don't create variables just to return them.  Do `return InsertAt(input, number, index);` instead.

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+    /// <summary>
+    /// Removes the first element from the array.
+    /// </summary>
+    /// <param name="input">Source array.</param>
```
Minor nit: Now you're calling it source array which also doesn't make sense.  I get there is some difference "I am removing an element from...source" vs "I am adding an element to...destination".  But still, I think I'd just say "The array to remove an element from" or something like that.  I'm not sure source or destination are the right terms.

gregory-butler replied:
Writing comments is cumbersome but I think it's good practice to describe what I was trying to achieve. Thank you.

westonpace replied:
In general people don't comment every method they write.  Usually they separate their project (mentally or otherwise) into a "public" portion, methods that are meant to be called by other people that use your code, and a private portion, helper methods that help implement the public methods.  For every 1 public method (that you have to comment) you might have 10-50 private methods (that people usually don't comment, or they do so only briefly).

westonpace commented on ArrayOperations/Arrays.cs:
```diff
@@ -0,0 +1,324 @@
+        throw new ArgumentException("Value cannot be null or empty.", nameof(message));
+      }
+
+      Console.WriteLine(message);
```
Although it does shorten things I think having `message` here violates the "single responsibility principle" (more on that later) and so I'd keep...
```
Console.WriteLine(message);
Print(arr);
```
over trying to combine the two.

gregory-butler replied:
I understand. Thank you.
Does the 'PromptString' method from the 'LoginSimulator' violate SRP?

westonpace replied:
It's not always black&white.  I think PromptString is fine.

westonpace commented on ArrayOperations/Demo.cs:
```diff
@@ -0,0 +1,111 @@
+    {
+      Arrays.Print(input,"Initial array elements:");
+
+      const int numberToInsert = 12;
```
Use `NumberToInsert` (first letter capitalized) for constants.

gregory-butler replied:
By default Visual Studio and Rider recommend to use camel case and it stuck with me.

westonpace commented on LoginSimulator/Program.cs:
```diff
@@ -0,0 +1,56 @@
+    
+    private static string PromptString(string message, bool isKeyVisible = false)
+    {
+      if (string.IsNullOrWhiteSpace(message))
```
In general, don't do null checks for internal methods.  You will drive yourself crazy.

westonpace commented on LoginSimulator/Program.cs:
```diff
@@ -0,0 +1,56 @@
+      var output = "";
+      while (true)
+      {
+        var pressedKey = Console.ReadKey(!isKeyVisible);
```
Why not `Console.ReadLine`?

gregory-butler replied:
I chose to use  `Console.ReadKey` so I can hide the user input. From my findings this is the only way I could achieve this.

westonpace replied:
That's a cool trick and makes a lot of sense.

westonpace reviewed (commented):
Clearly you put a lot of work into this.  So I decided to be extra strict.  It looks great.

gregory-butler committed:
correct summary;
add <returns> tag.

gregory-butler committed:
remove check when array length is zero, user should be able to insert elements in an empty array.
correct summary;
remove comment for 'ArgumentException'.

gregory-butler committed:
remove redudant check for 'null' and zero length array.
correct summary;
remove comment for 'ArgumentException'.

gregory-butler committed:
replace exception throwing with return input statement.

gregory-butler committed:
merge implemententantion with declaration.

gregory-butler committed:
change constants case PascalCase.

gregory-butler committed:
refactor 'Print' method to respect SRP;
merge array length verfication from 'Sort' method in to one stament;
refactor method return whitout creating new variable;
replace comments with more meaninfull ones.

gregory-butler committed:
update method calling to allow make typed user name visible.

gregory-butler committed:
update method calling to allow typed user name to be visible.

gregory-butler commented:
Thank you very much for the detailed review. I really appreciate the feedback you have given me.
I updated my code and I will push the commits shortly.

gregory-butler committed:
Merge remote-tracking branch 'origin/Homework/2/ArrayOperations' into Homework/2/ArrayOperations

gregory-butler committed:
add 'Fluent Assertions'.

gregory-butler committed:
refactor:
- variables to constants;
- constants names to 'Pascal Case'.
- merge declaration with initialization.

gregory-butler committed:
implement fluent assertions.

gregory-butler committed:
refactor 'Print' method to return a formatted string;
rename 'Print' method to 'ToString'.

gregory-butler committed:
implement tests for 'ToString' method.
